wordpress-mobile/WordPress-iOS
# My Site Dashboard: handle failures and stale data

Part of #17873

Handle stale data and failure states. This PR also refactored how the ghost cells and error card is displayed.

https://user-images.githubusercontent.com/7040243/156432538-2071b5b2-4645-43ea-b425-3e629ed74fc0.mp4

### Additional info

I decided to remove all error/ghost logic related to the dashboard service and the view model. Instead, I created a singleton to manage the state of the current dashboard.

Using that, ghost and failure cards are able to know if they should present themselves. I don't like the idea of using a singleton for that but it looks like a better option than handling these cases manually in the service and the view model — as it was being done.

### To test

Remove the app from your simulator/device, then run it and enable the My Site Dashboard feature flag.

**Important**: if you just want to check the failure card or the ghost cell visuals, you can change the property for them on `DashboardCard` (just change it to `true`).

#### Ghost cells

1. Enable Network Link Conditioner and use the 100% Loss profile
2. Open the app
3. Tap Home
4. You should see the ghost cards being displayed

#### Error state when loading for the first time

1. Turn Network Link Conditioner off and disable your wi-fi
2. Run the app
3. Tap Home
4. At the bottom of the dashboard you should see a "Some data wasn't loaded" message
5. Pull to refresh
6. The message persists

#### Failure when updating

1. Turn the wi-fi on
2. Tap Home and load the dashboard for any of your sites
3. Turn wi-fi off
4. Close the app and re-run it
5. You should see a notice at the bottom with a message that the app couldn't update
5. You should **not** see the "Some data wasn't loaded" message
6. Pull to refresh
7. You should see the notice
8. Go to any other screen (or put the app in the background) and go back to the Home tab
9. The notice should be displayed again

## Regression Notes
1. Potential unintended areas of impact
n/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:
